src/components/Table/Table.tsx

```diff
@@ -1,27 +1,103 @@
-import { Table as MantineTable, TextInput } from "@mantine/core";
-import { type ChangeEvent, type ReactNode, useMemo, useState } from "react";
-import { IconSortAscending, IconSortDescending } from "@tabler/icons-react";
+import {
+  Table as MantineTable,
+  TextInput,
+  Pagination as MantinePagination,
+  createStyles,
+  Box,
+} from "@mantine/core";
+
+import {
+  type ChangeEvent,
+  type ReactNode,
+  useEffect,
+  useMemo,
+  useState,
+} from "react";
+
+import {
+  IconArrowBarToLeft,
+  IconArrowBarToRight,
+  IconArrowLeft,
+  IconArrowRight,
+  IconGripHorizontal,
+  IconSortAscending,
+  IconSortDescending,
+} from "@tabler/icons-react";
+
 import { omit } from "lodash/fp";
-import { filterData, getNextSortValue, sortData } from "../../utils/utils.ts";
+
+import {
+  filterData,
+  getMaximumPages,
+  getNextSortValue,
+  sortData,
+} from "../../utils/utils.ts";
+
+import { useElementSize } from "@mantine/hooks";
 
 export interface Column<TData> {
   columnName: string;
   exactMatch: boolean;
   header: string;
   cellRenderer?: (row: TData) => ReactNode;
+  isFilterable: boolean;
 }
 
 interface TableProps<TData> {
   data: TData[];
   columns: Array<Column<TData>>;
+  rowHeight: number;
 }
 
 export type Sort = Record<string, "asc" | "desc" | null>;
 export type Filter = Record<string, string>;
 
+const DEBOUNCE_DELAY = 20;
+const TABLE_PADDING = 50;
 export function Table<T extends { id: number }>(props: TableProps<T>) {
+  const { classes } = useStyles();
+
+  const { ref: tableRef, height: tableHeight } = useElementSize();
+  const { ref: paginationRef, height: paginationHeight } = useElementSize();
+
+  const [adjustedTableHeight, setTableHeight] = useState(tableHeight);
+  const [numberOfRows, setNumberOfRows] = useState<number>(
+    adjustedTableHeight / props.rowHeight
+  );
+
   const [filtersState, setFilters] = useState<Filter>({});
   const [sortState, setSorting] = useState<Sort>({});
+  const [pageIndex, setPageIndex] = useState<number>(1);
+
+  useEffect(() => {
+    setTableHeight(tableHeight);
+    setNumberOfRows(
```

Comment:
I think this state is redundant (`const [numberOfRows, setNumberOfRows]`), as `numberOfRows` is simply a derivative of the table height & row height. (remember the waterfall)

Same for the `tableHeight` state, why do we need it? It's literally returned as is from `useElementSize`, which you then store in state and cause an additional re-render for nor reason